libnavigation-core/api/current.txt

```diff
@@ -0,0 +1,443 @@
+  }
+
+}
+
+package com.mapbox.navigation.core.replay {
+
+  public final class MapboxReplayer {
+    ctor public MapboxReplayer();
+    method public void clearEvents();
+    method public double durationSeconds();
+    method public void finish();
+    method public void play();
+    method public void playFirstLocation();
+    method public void playbackSpeed(double scale);
+    method public com.mapbox.navigation.core.replay.MapboxReplayer pushEvents(java.util.List<? extends com.mapbox.navigation.core.replay.history.ReplayEventBase> events);
+    method public void pushRealLocation(android.content.Context context, double eventTimestamp);
+    method public void registerObserver(com.mapbox.navigation.core.replay.history.ReplayEventsObserver observer);
+    method public void seekTo(double replayTime);
+    method public void seekTo(com.mapbox.navigation.core.replay.history.ReplayEventBase replayEvent);
+    method public void stop();
+    method public boolean unregisterObserver(com.mapbox.navigation.core.replay.history.ReplayEventsObserver observer);
+    method public void unregisterObservers();
+  }
+
+  public final class ReplayLocationEngine implements com.mapbox.android.core.location.LocationEngine com.mapbox.navigation.core.replay.history.ReplayEventsObserver {
+    ctor public ReplayLocationEngine(com.mapbox.navigation.core.replay.MapboxReplayer mapboxReplayer);
+    method public void getLastLocation(com.mapbox.android.core.location.LocationEngineCallback<com.mapbox.android.core.location.LocationEngineResult> callback);
+    method public void removeLocationUpdates(com.mapbox.android.core.location.LocationEngineCallback<com.mapbox.android.core.location.LocationEngineResult> callback);
+    method public void removeLocationUpdates(android.app.PendingIntent? pendingIntent);
+    method public void replayEvents(java.util.List<? extends com.mapbox.navigation.core.replay.history.ReplayEventBase> replayEvents);
+    method public void requestLocationUpdates(com.mapbox.android.core.location.LocationEngineRequest request, com.mapbox.android.core.location.LocationEngineCallback<com.mapbox.android.core.location.LocationEngineResult> callback, android.os.Looper? looper);
+    method public void requestLocationUpdates(com.mapbox.android.core.location.LocationEngineRequest request, android.app.PendingIntent? pendingIntent);
+    field public static final com.mapbox.navigation.core.replay.ReplayLocationEngine.Companion! Companion;
+  }
+
+  public static final class ReplayLocationEngine.Companion {
+    method public int getInstances();
+    method public void setInstances(int p);
+    property public final int instances;
+  }
+
+  public final class ReplayLocationEngineKt {
+  }
+
+}
+
+package com.mapbox.navigation.core.replay.history {
+
+  public interface CustomEventMapper {
+    method public com.mapbox.navigation.core.replay.history.ReplayEventBase? map(String eventType, com.google.gson.internal.LinkedTreeMap<?,?> properties);
```

Comment:
`com.google.gson.internal.LinkedTreeMap<?,?> properties`

We shouldn't rely on / expose internal APIs to the public 😬 - if `LinkedTreeMap` changes without notice (it can because of internal package "contract") would make us break SemVer unnecessarily. That's the whole purpose of https://github.com/mapbox/mapbox-navigation-android/issues/2417

cc @kmadsen

Reply:
Yeah i also really dislike the gson.internal.LinkedTreeMap type, but it was the solution at the moment. 

Can think of a couple solutions
 - [ ] try new json serializer (moshi or kotlinx.serialisation)
 - [ ] remove GSon from ReplayHistoryMapper and CustomEventMapper interface

Reply:
ok maybe a baby step is simple https://github.com/mapbox/mapbox-navigation-android/pull/3140

Reply:
- [ ] if there's no any "public" `LinkedTreeMap` maintain a copy of it https://github.com/google/gson/blob/master/gson/src/main/java/com/google/gson/internal/LinkedTreeMap.java from our side

Reply:
it would be nice to not have a public reference to it. Also gson has a [bad reputation](https://blog.usejournal.com/beyond-gson-evaluating-json-parsers-for-android-kotlin-e7aa4bcc413e)

Reply:
> it would be nice to not have a public reference to it.

Agreed 👇 

> if there's no any "public" `LinkedTreeMap`

I should have said or alternative 😉